Add recipe for org-todoist

Lillenne opened the pull request:
### Brief summary of what the package does

An Emacs package for bidirectional incremental sync of org-mode elements with [Todoist](https://todoist.com/) using org exactly how you normally would in Emacs Org Mode.

### Direct link to the package repository

https://github.com/lillenne/org-todoist

### Your association with the package

Author/maintainer

### Relevant communications with the upstream package maintainer

None needed

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

riscy commented:
Thanks for this.   Have you seen https://melpa.org/#/todoist which also depends on org?

In the meantime, quick first pass -

⸺ `org-todoist.el` with byte-compile using Emacs 29.4:
```
org-todoist.el:174:21: Warning: reference to free variable `org-state'
org-todoist.el:192:53: Warning: reference to free variable `url-http-end-of-headers'
In org-todoist--do-sync-callback:
org-todoist.el:464:46: Warning: Unused lexical argument `open'
org-todoist.el:464:55: Warning: Unused lexical argument `old'
org-todoist.el:476:35: Warning: reference to free variable `url-http-end-of-headers'
In org-todoist--push:
org-todoist.el:770:17: Warning: Unused lexical variable `last-repeat'
org-todoist.el:776:17: Warning: Unused lexical variable `old-todo-kw'
In org-todoist--sort-by-child-order:
org-todoist.el:1456:42: Warning: Unused lexical argument `NODE'
org-todoist.el:1456:47: Warning: Unused lexical argument `PROPERTY'
org-todoist.el:1456:66: Warning: Unused lexical argument `TYPE'
In end of data:
org-todoist.el:225:46: Warning: the function `projectile-project-name' is not known to be defined.
```

⸺ `org-todoist.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- `;;; Commentary` is much wider than 80 characters
- org-todoist.el#L202: It's safer to sharp-quote function names; use `#'`
- org-todoist.el#L605: This double negation can be collapsed (`not` aliases `null`)
- org-todoist.el#L1314: Use `when ...` instead of `unless (null ...)`
- org-todoist.el#L202: Loading a package should rarely add hooks
- org-todoist.el#L324: Loading a package should rarely add hooks
- org-todoist.el#L45: Top-level `setq` should usually be replaced by `defvar` or `defconst`
- org-todoist.el#L1832: `concat` may be unneeded; `insert` can take multiple arguments
- org-todoist.el#L1833: `concat` may be unneeded; `insert` can take multiple arguments
```

⸺ Package and license:
- Note you have multiple licenses here -
  - pkg/org-todoist.el: MIT License
  - Repository: GNU General Public License v3.0

riscy commented:
Re: your license, I see what happened re: detecting an MIT license.  Nevertheless, org-todolist.el should have its own formal license boilerplate and/or an SPDX license identifier: https://spdx.dev/ids/

Lillenne commented:
Hi @riscy, thank you for the feedback! 

Regarding the other Todoist package: yes, I am a contributor on the other Todoist package but to fulfill my needs for the integration it had to be entirely rearchitected and I didn't want to take over the project / step on toes / have to get approvals for my changes. The gist of the differences are that other package is primarily a stateless Emacs API client using the Todoist REST API, whereas this project uses the Todoist Sync API and models the data as org elements, then uses the existing syntax tree and previous responses to send and receive incremental updates & support time-tracking, etc.

I addressed all of the comments above except for 
- `projectile-project-name` not being known to be defined: that branch of code is only reachable if projectile is available and after projectile has been loaded (via an `(if (require 'projectile nil 'no-error))`). I am using projectile here if the user has it installed (it is not a package requirement) to infer the Todoist project headline for org-capture.  What is the proper way to silence this message?
- Loading a package should rarely add hooks: it is difficult to track recurring task completion with the current model, since the TODO state is automatically reset and the LAST_REPEAT property says the completion time instead of the last due date, which is what is needed for the Todoist API. Currently, I hook into the `org-after-todo-state-change-hook` to determine when these are completed.
- Loading a package should rarely add hook: the other hook I have is automatically syncing with Todoist after an org-capture, similar to how org-gcal automatically syncs new calendar events. I do use Doom Emacs, so maybe this is set up via doom and not via the package. Is this an acceptable default, or should I instead remove the default and show how to add this in the documentation?

Thanks!

riscy commented:
Thank you for clarifying - I think you have a good alternative approach here.

> that branch of code is only reachable if projectile is available

You can hint the byte-compiler at this and remove the Warning with an `fboundp` check, e.g.
```elisp
          (if (and org-todoist-infer-project-for-capture
                   (require 'projectile nil 'no-error))
              (if (fboundp 'projectile-project-name)
                  (if (string= (projectile-project-name) "-")
                      (completing-read "Which project? " project-names)
                    (projectile-project-name)))
```
I'm also seeing:
```
org-todoist.el:778:17: Warning: Unused lexical variable `is-recurring'
org-todoist.el:781:17: Warning: Unused lexical variable `lr'
```
but if these are truly meant to be unused you can hint the compiler by prefixing with an underscore, e.g. `_is_recurring`.

> Currently, I hook into the org-after-todo-state-change-hook to determine when these are completed.

It's not the hook itself, but the fact that it occurs globally and at the top level.  This means if someone loads your package - even just to depend on some part of it as a library - the hooks take effect in every single org buffer.  If it's possible (and I admit it is sometimes not possible) it's better to add these hooks when they're needed, and to make them LOCAL to the buffer where they're needed (`add-hook` has a LOCAL arg).

riscy commented:
Friendly ping. :)

Lillenne commented:
I've been meaning to get back to cleaning this up for a very long time! It has been a busy year and then Todoist deprecated the API version I was using. I've updated the package and added a _lot_ of new features. Package lint and checkdoc pass and the file now byte-compiles cleanly. 

I introduced a minor mode and made _most_ hooks buffer-local. The sync after org capture hook is added by default when loading the package and is not buffer-local. Additionally, a hook is added to org-mode to turn on the minor mode automatically when visiting the `org-todoist-file`. If  `'auto-minor-mode-alist` is bound, this hook is not added and the package instead hooks into auto-minor-mode with the current `org-todoist-file`. 

If there is anything else needed, please let me know!

Lillenne commented:
Sorry, I did the checks and then quickly fixed a bug and ruined it with the documentation! I must be doing the byte compilation step incorrectly, because it is not showing the unused lexical variable issue? I added a line to `(ignore <the var>)` and fixed the documentation again. Thank you for your patience and sorry for the extra work!

riscy commented:
> because it is not showing the unused lexical variable issue?

`org-export-show-temporary-export-buffer` is defined in the builtin `ox`, so you need to add `(require 'ox)` at the top of the file - that should fix it.

Here are some other issues following the recent code change -

- org-todoist.el#L575: Loading a package should rarely add hooks (how does a user undo this?  Can we gate this behind a mode?)
- org-todoist.el#L2089: Consider using %F instead of %Y-%m-%d in time strings (optional)
- org-todoist.el#L2093: Consider using %F instead of %Y-%m-%d in time strings
- org-todoist.el#L2089: Consider using %R instead of %H:%M in time strings
- org-todoist.el#L2091: Consider using %R instead of %H:%M in time strings
- org-todoist.el#L2093: Consider using %R instead of %H:%M in time strings
